src/docs/truffle/advanced/event-system.md

```diff
@@ -0,0 +1,418 @@
+match `"fetchSolcList:succeed"` and `"unbox:preparingToDownload:succeed"`.
+In this way you are able to write handlers that match against batches of events.
+
+The event matcher string must be a key in the `handlers` object whose value
+must be an array of functions. These functions will be executed when that
+event matcher matches an emitted event. Every time an event matcher matches
+an emitted event, each of its functions will be executed. Here is a
+simple "hello world" example:
+
+```javascript
+module.exports = {
+  handlers: {
+    "compile:start": [
+      function() {
+        console.log("hello world!");
+      }
+    ]
+  }
+};
+```
+
+In the above example, every time the `"compile:start"` event is emitted,
+`"hello world!"` will be logged to the console.
+
+*NOTE: Currently you must use `function` syntax when creating the handler
+functions in order to have the appropriate `this` value. You can use arrow
+functions in your handler functions, but you will lose the `this` reference
+to the Subscriber if you do so.*
+
+## More on subscribers
+
+When you are creating your handlers, you will have access to Subscriber class
+methods. Most of these shouldn't be used directly except for the
+`removeListener` method. If at some point you need to remove a listener that
+was added, you can call `this.removeListener(<eventMatcher>)` to "detach"
+the handlers listed under that specific event matcher.
+
+You will also have access to everything that you made references to in the
+initialization function. For example: if I wanted to use an external library,
+perhaps one for colorful logging, I would require it in my JavaScript and
+create a reference to it in the initialization like so:
+
+```javascript
+const colors = require("colors");
+
+module.exports = {
+  initialization: function() {
+    this.colors = colors;
+  },
+  // ......
+};
+```
+
+I would then be able to access `this.colors` in my handlers property! So
+extending the example above could yield the following code:
+
+```javascript
+const colors = require("colors");
+
+module.exports = {
+  initialization: function() {
+    this.colors = colors;
+  },
+  handlers: {
+    "compile:compiledSources": [
+      function(data) {
+        const { sourceFileNames } = data;
+        const message = this.colors.rainbow(`The source files are ${sourceFileNames}`);
+        console.log(message);
+      }
+    ]
+  }
+};
+```
+
+This would log the source filenames in rainbow colors whenever the
+`"compile:compiledSources"` event is emitted. Remember, some events (not all)
+provide data to your handlers at the time of execution. The
+`"compile:compiledSources"` provides an array of all source filenames to this
+particular handler. Check the chart below to see which events are provided
+with what data.
+
+*Reminder: The `this` in your Subscriber files refers to the Subscriber class
+that is instantiated from the file you create and not the `this` in the file.*
+
+## Currently supported events
```

Comment:
We should either move this section to the "Reference" area of the Truffle docs, or rethink the Reference section entirely (leaning toward the second one). I'm thinking out loud here--for now LGTM as is!